Various fixes for Ninja support

SebastianBoe opened the pull request:
When building for Ninja we were accidentally using the wrong depfile
in the second pass. This commit moves the LINKER_SCRIPT_DEP logic into
the custom command function so that it can be linker-pass-aware and
set the depfile appropriately.

This should fix an issue where Ninja reported:

ninja: error: expected depfile 'zephyr/linker.cmd.dep' to mention
'zephyr/linker_pass2.cmd', got 'zephyr/linker.cmd'

But this has not been reproduced. It has however been confirmed that a
dependency issue with linker_pass2.cmd has been fixed because ninja no
longer regenerates linker_pass2.cmd on every build like this:

$ ninja
[1/79] Generating always_rebuild
Building for board qemu_x86
[2/3] Generating linker_pass2.cmd
[3/3] Linking C executable zephyr/zephyr.elf

Now:

$ ninja
[1/78] Generating always_rebuild
Building for board qemu_x86

SebastianBoe committed:
cmake: ld: Refactor the linker script's cpp command construction

Construct the custom command for preprocessing the linker script with
a function to avoid copy-paste errors between linker passes.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
cmake: Ninja: ld: Use the correct depfile in the second pass

When building for Ninja we were accidentally using the wrong depfile
in the second pass. This commit moves the LINKER_SCRIPT_DEP logic into
the custom command function so that it can be linker-pass-aware and
set the depfile appropriately.

This should fix an issue where Ninja reported:

ninja: error: expected depfile 'zephyr/linker.cmd.dep' to mention
'zephyr/linker_pass2.cmd', got 'zephyr/linker.cmd'

But this has not been reproduced. It has however been confirmed that a
dependency issue with linker_pass2.cmd has been fixed because ninja no
longer regenerates linker_pass2.cmd on every build like this:

$ ninja
[1/79] Generating always_rebuild
Building for board qemu_x86
[2/3] Generating linker_pass2.cmd
[3/3] Linking C executable zephyr/zephyr.elf

Now:

$ ninja
[1/78] Generating always_rebuild
Building for board qemu_x86

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

nashif commented:
can you try this change on top of #5522 and run sanitycheck with

`# sanitycheck -p qemu_x86 -N`

I am getting lots of failures. for example:

```
FAILED: zephyr/zephyr_prebuilt.elf
: && ccache /home/nashif/Work/SDK/0.9.2/sysroots/x86_64-pokysdk-linux/usr/bin/i586-zephyr-elf/i586-zephyr-elf-gcc    zephyr/CMakeFiles/zephyr_prebuilt.dir/misc/empty_file.c.obj  -o zephyr/zephyr_prebuilt.elf  -T zephyr/linker.cmd -Wl,-Map=/home/nashif/Work/zephyr/samples/net/coap_server/build/zephyr/zephyr.map -u_OffsetAbsSyms -u_ConfigAbsSyms -e__start -Wl,--start-group -Wl,--whole-archive libapp.a zephyr/libzephyr.a zephyr/subsys/bluetooth/common/libsubsys__bluetooth__common.a zephyr/subsys/bluetooth/host/libsubsys__bluetooth__host.a zephyr/subsys/net/libsubsys__net.a zephyr/subsys/net/ip/libsubsys__net__ip.a zephyr/subsys/net/ip/l2/libsubsys__net__ip__l2.a zephyr/arch/x86/soc/ia32/libarch__x86__soc__ia32.a -Wl,--no-whole-archive zephyr/kernel/libkernel.a zephyr/CMakeFiles/offsets.dir/arch/x86/core/offsets/offsets.c.obj -Wl,--end-group -L"/home/nashif/Work/SDK/0.9.2/sysroots/i586-zephyr-elf/usr/lib/i586-zephyr-elf/6.2.0" -L/home/nashif/Work/zephyr/samples/net/coap_server/build/zephyr -lgcc -Wl,--print-memory-usage -nostartfiles -nodefaultlibs -nostdlib -static -no-pie -Wl,-X -Wl,-N -Wl,--gc-sections -Wl,--build-id=none && :
Memory region         Used Size  Region Size  %age Used
             ROM:      138843 B      4092 KB      3.31%
             RAM:        108 KB         4 MB      2.64%
        IDT_LIST:         492 B         2 KB     24.02%
        MMU_LIST:        9296 B         1 KB    907.81/home/nashif/Work/SDK/0.9.2/sysroots/x86_64-pokysdk-linux/usr/libexec/i586-zephyr-elf/gcc/i586-zephyr-elf/6.2.0/real-ld: zephyr/zephyr_prebuilt.elf section `.noinit."/home/nashif/Work/zephyr/subsys/bluetooth/host/hci_core.c".2' will not fit in region `MMU_LIST'
/home/nashif/Work/SDK/0.9.2/sysroots/x86_64-pokysdk-linux/usr/libexec/i586-zephyr-elf/gcc/i586-zephyr-elf/6.2.0/real-ld: region `MMU_LIST' overflowed by 8272 bytes
collect2: error: ld returned 1 exit status
%
ninja: build stopped: subcommand failed.
```

SebastianBoe commented:
Will do.

SebastianBoe commented:
@nashif : Those failures are due to the first point in this issue: https://github.com/zephyrproject-rtos/zephyr/issues/5334

They are unrelated to this patch, so this patch could (should) be applied before it is fixed.

Will start working on the linker issue now.

SebastianBoe committed:
cmake: ninja: Change how CMake names the kernelspace archives

The linker script places kernelspace and userspace archives in
different sections. But the linker script itself does not determine
what archives are in what space, that is done by CMake.

CMake passes the list of kernelspace archives to the linker script
through defines, like this:

-DNUM_KERNEL_OBJECT_FILES=3
-DKERNEL_OBJECT_FILE_0=path/to/archive_a.a
-DKERNEL_OBJECT_FILE_1=path/to/archive_b.a
-DKERNEL_OBJECT_FILE_2=path/to/archive_c.a

These paths are relative, and since Ninja and Make invoke the linker
with different "working directories"[0], the relative paths need to be
different. This patch rectifies the relative path when using Ninja.

This fixes #5343

[0] https://gitlab.kitware.com/cmake/cmake/issues/17448

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe commented:
The failure is a false-positive checkpatch error. The patch is actually good (As far as CI is concerned).

SebastianBoe committed:
cmake: Ninja: Fix the external_lib sample

This samples was failing to build with Ninja because Ninja detected
that libmylib.a was missing. Adding it as a BYPRODUCT in the
ExternalProject fixes the issue.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
cmake: LD: Add comment explaining the -P flag

Add a comment that explains what the -P flag does. The explanatory
comment is useful when you want to find the flag so you can comment it
out while debugging the link stage.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>
